lib/semantics/check-allocate.cc

```diff
@@ -0,0 +1,411 @@
+                        .declTypeSpec;
+  }
+
+  const parser::Expr *parserSourceExpr{nullptr};
+  for (const parser::AllocOpt &allocOpt :
+      std::get<std::list<parser::AllocOpt>>(allocateStmt.t)) {
+    std::visit(
+        common::visitors{
+            [&](const parser::StatOrErrmsg &statOrErr) {
+              std::visit(
+                  common::visitors{
+                      [&](const parser::StatVariable &statVariable) {
+                        analyzer.Analyze(statVariable.v);
+                        if (info.gotStat) {  // C943
+                          context.Say(
+                              "STAT may not be duplicated in a ALLOCATE statement"_err_en_US);
+                        }
+                        info.gotStat = true;
+                      },
+                      [&](const parser::MsgVariable &msgVariable) {
+                        analyzer.Analyze(msgVariable.v);
+                        if (info.gotMsg) {  // C943
+                          context.Say(
+                              "ERRMSG may not be duplicated in a ALLOCATE statement"_err_en_US);
+                        }
+                        info.gotMsg = true;
+                      },
+                  },
+                  statOrErr.u);
+            },
+            [&](const parser::AllocOpt::Source &source) {
+              analyzer.Analyze(source.v);
+              if (info.gotSrc) {  // C943
+                context.Say(
+                    "SOURCE may not be duplicated in a ALLOCATE statement"_err_en_US);
+                stopCheckingAllocate = true;
+              }
+              if (info.gotMold || info.gotTypeSpec) {  // C944
+                context.Say(
+                    "At most one of source-expr and type-spec may appear in a ALLOCATE statement"_err_en_US);
+                stopCheckingAllocate = true;
+              }
+              parserSourceExpr = &source.v.value();
+              info.gotSrc = true;
+            },
+            [&](const parser::AllocOpt::Mold &mold) {
+              analyzer.Analyze(mold.v);
+              if (info.gotMold) {  // C943
+                context.Say(
+                    "MOLD may not be duplicated in a ALLOCATE statement"_err_en_US);
+                stopCheckingAllocate = true;
+              }
+              if (info.gotSrc || info.gotTypeSpec) {  // C944
+                context.Say(
+                    "At most one of source-expr and type-spec may appear in a ALLOCATE statement"_err_en_US);
+                stopCheckingAllocate = true;
+              }
+              parserSourceExpr = &mold.v.value();
+              info.gotMold = true;
+            },
+        },
+        allocOpt.u);
+  }
+
+  if (stopCheckingAllocate) {
+    return std::nullopt;
+  }
+
+  if (info.gotSrc || info.gotMold) {
+    CHECK(parserSourceExpr);
+    CHECK(parserSourceExpr->typedExpr);  // TODO: Can we reach this spot without
+                                         // a valid expression?
```

Comment:
As things stand you are correct, if the expression is invalid ExprChecker would have found the error and this checker would never be called. I want to change it so that the other checkers are called even if there are expression errors. So it's better to check that typedExpr is present.

Reply:
Thanks for the heads up on PR #422. I modified this part to use `GetExpr`. As I understood from PR #424, if `GetExpr` is returning a null pointer, it means that the expression was analyzed but there was an error reported and allocate tests should stop without emitting new errors regarding this. Is this correct?

Reply:
Right